src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,693 @@
+    bool create_missing_root_path)
+    : IStorage(table_id)
+    , WithContext(context_->getGlobalContext())
+    , root_path(zkutil::extractZooKeeperPath(root_path_, false))
+    , primary_key(primary_key_)
+    , zookeeper_name(zkutil::extractZooKeeperName(root_path_))
+    , log(&Poco::Logger::get("StorageKeeperMap"))
+{
+    setInMemoryMetadata(metadata);
+
+    if (root_path.empty())
+        throw Exception("root_path should not be empty", ErrorCodes::BAD_ARGUMENTS);
+    if (!root_path.starts_with('/'))
+        throw Exception("root_path should start with '/'", ErrorCodes::BAD_ARGUMENTS);
+
+    std::filesystem::path root_path_fs{root_path};
+    auto metadata_path_fs = root_path_fs / "ch_metadata";
+    metadata_path = metadata_path_fs;
+    lock_path = metadata_path_fs / "lock";
+    tables_path = metadata_path_fs / "tables";
+    dropped_path = metadata_path_fs / "dropped";
+
+    if (attach)
+    {
+        checkTable<false>();
+        return;
+    }
+
+    auto client = getClient();
+
+    if (root_path != "/" && !client->exists(root_path))
+    {
+        if (!create_missing_root_path)
+        {
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Path '{}' doesn't exist. Please create it or set 'create_missing_root_path' to true'",
+                root_path_);
+        }
+        else
+        {
+            LOG_TRACE(log, "Creating root path {}", root_path);
+
+            size_t cur_pos = 0;
+            do
+            {
+                size_t search_start = cur_pos + 1;
+                cur_pos = root_path.find('/', search_start);
+                if (search_start == cur_pos)
+                    throw Exception("root_path is invalid, contains subsequent '/'", ErrorCodes::BAD_ARGUMENTS);
+
+                auto path = root_path.substr(0, cur_pos);
+                client->createIfNotExists(path, "");
+            } while (cur_pos != std::string_view::npos);
+        }
+    }
+
+
+    ZooKeeperLock lock(lock_path, client, true);
+
+    bool locked = false;
+    try
+    {
+        lock.lock();
+        locked = true;
+    }
+    catch (const Coordination::Exception & e)
+    {
+        if (e.code != Coordination::Error::ZNONODE)
+            throw;
+    }
+
+    {
+        Coordination::Stat stats;
+        auto exists = client->exists(root_path, &stats);
+        if (!exists)
+            throw Exception(DB::ErrorCodes::LOGICAL_ERROR, "Path '{}' should exist, but was deleted from another source", root_path);
+    }
+
+
+    const auto set_tables_num = [&]
+    {
+        std::string tables_num_str;
+        auto exists = client->tryGet(tables_path, tables_num_str);
+
+        if (!exists)
+        {
+            client->create(tables_path, toString(1), zkutil::CreateMode::Persistent);
+        }
+        else
+        {
+            uint64_t tables_num = parse<uint64_t>(tables_num_str);
+            client->set(tables_path, toString(tables_num + 1), zkutil::CreateMode::Persistent);
```

Comment:
```suggestion
            client->set(tables_path, toString(tables_num + 1));
```
zkutil::CreateMode::Persistent coincidentally matches the default parameter, causes the version parameter to be passed as 0.